Add support for NXP S32 Watchdog driver

quangbuitrong opened the pull request:
This PR introduces support for NXP S32 RTU.SWT (WATCHDOG) driver and enable its usage for `s32z270dc2_r52` boards.

The initial version only supports interrupt triggering when timeout expires and does not support the reset function because currently, the board does not support running on flash.

Cherry-picked 924715be84b02117603df8304bcc51e1684f9c6b commit from NXP S32 System Timer Module Pull Request (https://github.com/zephyrproject-rtos/zephyr/pull/51451) to get the RTU index for the purpose of getting the correct instance of the Watchdog interrupts.

A simple test case to test watchdog interrupt can be triggered when the timeout expires is also added for this board.

quangbuitrong committed:
drivers: watchdog: add support for NXP S32 S32ZE

 Add driver shim for the NXP S32 SWT module.

Signed-off-by: Quang Bui Trong <quang.buitrong@nxp.com>

quangbuitrong committed:
boards: arm: s32z270dc2_r52: enable watchdog support

 Enable RTU.SWT (Real-Time Unit.Software Watchdog Timer) instances on
 s32z270dc2_r52 boards. Module clock frequency is fixed to 48 Mhz.

Signed-off-by: Quang Bui Trong <quang.buitrong@nxp.com>

quangbuitrong committed:
boards: s32z270dc2_r52: add docs for watchdog shim driver

 Document watchdog shim driver supported features for s32z270dc2_r52
 boards.

Signed-off-by: Quang Bui Trong <quang.buitrong@nxp.com>

danieldegrasse commented on tests/boards/nxp_s32/swt/src/main.c:
```diff
@@ -0,0 +1,86 @@
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/watchdog.h>
+
+#define DEVICE_FOR_DT_COMPAT(n, compat) DEVICE_DT_GET(DT_INST(n, compat)),
```
is there anything about this test that is board specific? Or could a new test be added to `tests/drivers/watchdog`, that verifies the watchdog callback feature without resetting the SOC?

manuargue replied:
+1 with the idea of extending existing tests / creating new generic test, for this use-case

quangbuitrong replied:
This test isn't exactly board specific. Yes, it verifies the watchdog callback feature without resetting the SOC.
I can put it in `tests/drivers/watchdog` but I'm not sure whether vendors need this feature or not. So I put it in `tests/boards`. Also, this test isn't necessary when `s32z270dc2_r52 ` board supports running from flash.

danieldegrasse replied:
I think unless the test needs to be specific to a board, it should be placed in `tests/drivers/watchdog`. This will at least make it possible for other vendors to use the test if desired.

manuargue commented on drivers/watchdog/wdt_nxp_s32.c:
```diff
@@ -0,0 +1,208 @@
+			 &swt_nxp_s32_driver_api);
+
+
+#if DT_NODE_HAS_STATUS(SWT_NODE(0), okay)
```
The NXP S32 HAL is index-based, already discussed in https://github.com/zephyrproject-rtos/zephyr/pull/49161#discussion_r949770066. I will address this later as a larger PR for all drivers supported so far.

danieldegrasse replied:
Thanks for the clarification

danieldegrasse reviewed (changes requested):
Please change the board specific watchdog test to be generic, and include it in `tests/drivers/watchdog`. Even if this test is not currently used for other vendors, I can see it being valuable for platforms that execute from RAM

quangbuitrong committed:
tests: drivers: watchdog: add test watchdog for s32z270dc2_r52

 This test to verify Watchdog install/setup/feed can work and
 interrupt can be triggered on timeouts. This test doesn't test the
 case of reseting the board on time out because flash isn't currently
 support.

Signed-off-by: Quang Bui Trong <quang.buitrong@nxp.com>

quangbuitrong commented:
I created a new generic test for watchdog: `tests/drivers/watchdog/wdt_basic_reset_none`, this test supports boards running from RAM and cannot reset soc when watchdog expires timeout.

danieldegrasse commented on boards/arm/s32z270dc2_r52/s32z270dc2_r52.dtsi:
```diff
@@ -68,3 +68,23 @@
 };
+
+&swt0 {
+	clock-frequency = <48000000>;
```
Should one of the watchdog timers be `status=okay`, so that it can be configured by applications? It seems the watchdog timers are currently only being set to `status=okay` in tests that utilize them

manuargue reviewed (commented):
LGTM. Pls fix conflicts

danieldegrasse commented:
@quangbuitrong https://github.com/zephyrproject-rtos/hal_nxp/pull/209 has been merged. However, I would wait for https://github.com/zephyrproject-rtos/zephyr/pull/53229 to merge before updating this PR, as otherwise you'll need to rebase twice
